### ISSUE TITLE
Labeling algorithms (e.g., occlusion)?

### ISSUE THREAD

**Issue by mbostock** [observablehq/plot#27]
In some line plots, if the cardinality of z is relatively low, it would be nice to label the lines say at the start or end. Similarly for dot plots, it would be lovely to have an automatic labeling strategy. Related [vega-label](https://github.com/vega/vega/tree/master/packages/vega-label).

**Comment by Fil** [comment]
I also like the "white stroke clone" strategy from https://observablehq.com/@mbostock/inequality-in-american-cities and think it should be available from Plot.text (and maybe occlusion as well!).

**Comment by Fil** [comment]
Here's a prototype for a labelling strategy based on @fil/occlusion
https://observablehq.com/@observablehq/plot-occlusion-27

![](https://static.observableusercontent.com/thumbnail/fc454ff7c630782f2bac4488a40dfba1f89b897e561b2b0040fbd9636e679251.jpg)

On this scatterplot, the occludeText function is able to avoid putting text on top of symbols. It's still not capable of finding a correct position to put the texts. It also needs to wait for the chart to be appended to the DOM before it can kick in, as it needs to measure the actual bounding boxes.

**Comment by mbostock** [comment]
Can we avoid getBoundingClientRect by using the same metrics approach that Plot.text uses?

**Comment by Fil** [comment]
Yes it's one of the todos. Some challenges (probably not too hard):
* we'll have to compute intersections not only for text but for symbols, rects, lines, paths… (this might also be useful for interactions);
* this is yet another case where it will be useful to know in which facet we are, because the same position in two facets is not creating an occlusion.

**Comment by Fil** [comment]
Voronoi centroid initializer https://observablehq.com/@fil/plot-voronoi-labels

**Comment by Fil** [comment]
A typical ask is to limit occlusion of end-of-line labels, like in https://observablehq.com/@d3/cancer-survival-rates (unfinished Plot version: https://observablehq.com/@observablehq/plot-cancer-survival-rates)

**Comment by Fil** [comment]
See also mapbox’s text-variable-anchor property
https://docs.mapbox.com/mapbox-gl-js/example/variable-label-placement/

**Comment by Fil** [comment]
Another example, in which the temptation was to use dodgeY (but that's not how it works) to avoid label occlusion
https://observablehq.com/d/1eb96ffa035f4676

**Comment by huw** [comment]
Here's a hack for occluding axis ticks by combining `dodge` with a filter initializer: https://observablehq.com/@huw/occlude-axis (FWIW it would be good if any occlusion transform also supported axes)
